### PR TITLE
[android] zoom buttons are not disabled fixed

### PR DESCRIPTION
**Pull request by Aryan-Sagavekar**
[Android][Regression]
Fixed issue [#4680](https://github.com/organicmaps/organicmaps/issues/4680) , where the zoom buttons did not disable from the map. Also the earlier bug [#4458](https://github.com/organicmaps/organicmaps/pull/4458)  is not reintroduced.

### CODE REVIEW

File: android/src/app/organicmaps/maplayer/MapButtonsController.java [hunk 1]
```diff
@@ -268,10 +268,14 @@ public float getBottomButtonsHeight()
...
   {
-    if (show)
+    if (show) {
       UiUtils.show(mFrame);
-    else
+      showButton(true, MapButtons.zoom);
+    }
+    else {
       UiUtils.hide(mFrame);
+      showButton(false, MapButtons.zoom);
```
Comment (arnaudvergnet) [inline android/src/app/organicmaps/maplayer/MapButtonsController.java:277]: The zoom buttons are inside the frame, so there is no need to hide them if we hide the parent frame.

**Comment by arnaudvergnet** [comment]
Thanks for the PR, please add the DCO to your commit as described [here](https://github.com/organicmaps/organicmaps/pull/4689/checks?check_run_id=11763043209)

File: android/src/app/organicmaps/maplayer/MapButtonsController.java (hunk 1 above)
Comment (arnaudvergnet) [inline android/src/app/organicmaps/maplayer/MapButtonsController.java:277]: The zoom buttons are inside the frame, so there is no need to hide them if we hide the parent frame.

**Review by arnaudvergnet — CHANGES_REQUESTED** [review]

File: android/src/app/organicmaps/maplayer/MapButtonsController.java (hunk 1 above)
Comment (Aryan-Sagavekar) [reply android/src/app/organicmaps/maplayer/MapButtonsController.java:277]: if that line is not added then the zoom buttons pop up again when we enter the edit mode.

File: android/src/app/organicmaps/maplayer/MapButtonsController.java [hunk 2]
```diff
@@ -268,10 +268,14 @@ public float getBottomButtonsHeight()
 
   public void showMapButtons(boolean show)
   {
-    if (show)
+    if (show) {
       UiUtils.show(mFrame);
```
Comment (biodranik) [inline android/src/app/organicmaps/maplayer/MapButtonsController.java:272]: Please read our style guide and see how the surrounding code is formatted.

```suggestion
    if (show)
    {
      UiUtils.show(mFrame);
```

File: android/src/app/organicmaps/maplayer/MapButtonsController.java [hunk 3]
```diff
@@ -268,10 +268,14 @@ public float getBottomButtonsHeight()
...
-    else
+      showButton(true, MapButtons.zoom);
+    }
+    else {
       UiUtils.hide(mFrame);
```
Comment (biodranik) [inline android/src/app/organicmaps/maplayer/MapButtonsController.java:276]:
```suggestion
    else
    {
      UiUtils.hide(mFrame);
```

File: android/src/app/organicmaps/maplayer/MapButtonsController.java (hunk 1 above)
Comment (biodranik) [reply android/src/app/organicmaps/maplayer/MapButtonsController.java:277]: Why do zoom buttons pop up again if this line is not added if the frame is hidden? Is the frame displayed again for some reason in the edit mode? What causes it? How can we fix the root cause?

Comment (Aryan-Sagavekar) [reply android/src/app/organicmaps/maplayer/MapButtonsController.java:277]: This PR might be a temporary solution. But I will certainly try to find the root cause.

Comment (biodranik) [reply android/src/app/organicmaps/maplayer/MapButtonsController.java:278]: Temporary solutions easily stack on each other and then lead to other unexpected issues. Let's focus on fixing the root cause.

Comment (Aryan-Sagavekar) [reply android/src/app/organicmaps/maplayer/MapButtonsController.java:278]: So shall I close this pull request then?

Comment (biodranik) [reply android/src/app/organicmaps/maplayer/MapButtonsController.java:278]: It would be great if you find a root cause, and fix it in this PR, without creating another one.

**Comment by vng** [comment]
Well, I'd also expect that showing/hiding parent mFrame only is enough here. And I don't understand why it's not working now.

But I don't like the current iOS behavior when hiding +/- buttons cause layout recalculation, and the position button jumps up from the usual familiar position.

@biodranik @arnaudvergnet 
Should we reconsider it also in Android? Like keep a transparent Frame (for constant layout) but show/hide buttons only?

File: android/src/app/organicmaps/maplayer/MapButtonsController.java (hunk 1 above)
Comment (Aryan-Sagavekar) [reply android/src/app/organicmaps/maplayer/MapButtonsController.java:278]: The zoom buttons are included in the mainframe and disappear and reappear as other buttons present there.
The thing is zoom buttons have a separate setting of disabling/enabling them unlike the other mainframe buttons, this is causing them to act differently and needs to add those statements there.
I'll try to find a different approach if this doesnt sound right to you.

**Comment by biodranik** [comment]
@vng My position button should definitely not jump when these buttons are hidden.

**Comment by vng** [comment]
In this case, the solution to hide buttons only (not mFrame) is ok. Should check how it works with touches ..

**Comment by vng** [comment]
I suspect that the current mFrame is useless because I don't have any other ideas, why it's not hiding now :)

**Comment by arnaudvergnet** [comment]
For the moment, the buttons can be show/hidden all at the same time when a user presses in an empty map area to go fullscreen. This is done by the `showMapButtons` function, which shows/hides the buttons frame instead of buttons individually.

Individual buttons can be hidden when the place page slides. If a button is too close to the compass button, then it is hidden.

The issue is that when the activity is resumed (going back from settings/editor), the `showMapButtons` is called again to resume to the last mode (fullscreen  or normal).

As the zoom buttons follow a setting the user can change, we need to update its state when resuming, so putting the logic as this PR does in the `showMapButtons` function makes sense. But if the place page was already extended enough so the zoom buttons are under the compass, then this call will show them again.

We need to implement a specific logic to "resume" the state of the zoom buttons. It should be a new function that checks the position of the zoom buttons against the position of the place page. So instead of calling `showButton(false, MapButtons.zoom)`, we could call something like `resumeZoomButtons()`.

**Comment by Aryan-Sagavekar** [comment]
@arnaudvergnet as you said I created the resumeZoomButtons() function and it works as expected. If you want you can try it.
image
image
Let me know if there are any changes I should make.
Also showButton() function only has one use, maybe that could be changed.

**Comment by arnaudvergnet** [comment]
From looking at the code, it seems your solution will bring back #4458.

The function `resumeZoomButtons` should check the position of the buttons to see if they are not under the compass. For this to work, we need to save the current place page translation (updated in `applyMove`) as a class member, and call `updateButtonsVisibility` with that value inside `resumeZoomButtons`. `updateButtonsVisibility` uses `showButton` so it will handle config changes.

**Comment by Aryan-Sagavekar** [comment]
It's not bringing back #4458 but i'll make the changes as you suggested.

**Comment by Aryan-Sagavekar** [comment]
image
@arnaudvergnet like this? I am confused as to where to call that function as it creates issues when called in showMapButtons().

**Comment by Aryan-Sagavekar** [comment]
hey @arnaudvergnet,
Something like this?
image
It works well until you go to landscape mode.

**Comment by biodranik** [comment]
Please focus on the simplest and reliable fix for all possible use cases and modes.

**Comment by Aryan-Sagavekar** [comment]
The previous logic was working. Here's a screen recording as a proof. 

https://user-images.githubusercontent.com/109228835/224492345-ce0a6b53-5322-4df1-980a-84942869de6f.mp4

so is it ok to use the logic shown in this https://github.com/organicmaps/organicmaps/pull/4689#issuecomment-1459832375

**Comment by arnaudvergnet** [comment]
Your video does not show the issue fixed in https://github.com/organicmaps/organicmaps/pull/4458 is not reintroduced. And after testing your code, I can confirm it is reintroduced. To reproduce it, you need to open the place page so the zoom buttons go under the compass and disappear. Once in that state, click on "Edit place" and go back: you will see the zoom buttons reappeared under the compass when they should not.

Here is an example (recorded with your changes):

https://user-images.githubusercontent.com/80701113/224504266-aedeb1f7-1801-46a1-9ba4-4b085be3dd14.mp4



**I looked a bit more into this and here is how we can reliably fix this:**

There is already the function `updateButtonsVisibility()` available which does what we need: it loops over all buttons and checks if they are under the compass to know if they should be shown or hidden. It is not used (forgot in a previous refactoring) and to make it properly work we need to check if `mInnerLeftButtonsFrame` and `mInnerRightButtonsFrame` are not null.

You can call that function in `showMapButtons()` as a replacement of the `resumeMapButtons()` function you created.

There's only one more change needed for everything to work: in `updateButtonsVisibility(final float translation, @Nullable View parent)` we need to compare the returned value from `getViewTopOffset` using `>=` and not `>` because on the first call the frame translation and the distance to top of the buttons may both be 0.

And please, do not post screenshots of your code, just update the PR even if you are not sure, it is easier for us to review and test your changes.

Please tell me if you have trouble implementing the fix! It is a bit complex because the `MwmActivity` is interacting directly with the fragment, so the `showButtons` function may be called while the `MapButtonsController` view is not yet ready. This needs to be cleaned up but it will be the work of another PR.

**Comment by Aryan-Sagavekar** [comment]
Thank you sir,
I'll make sure it is fixed this time

**Comment by arnaudvergnet** [comment]
No pressure, thanks a lot for investing your time in the project!

**Comment by Aryan-Sagavekar** [comment]
Did some minor changes to the updateButtonsVisibility() and resumeZoomButtons() and finally it worked. Tested it as your video did to check whether they pop up again and they did not.

File: android/src/app/organicmaps/maplayer/MapButtonsController.java [hunk 4]
```diff
@@ -369,4 +373,10 @@ public void onLayoutChange(View v, int left, int top, int right, int bottom, int
       mContentView.removeOnLayoutChangeListener(this);
     }
   }
-}
+
+  public void resumeZoomButtons()
```
Comment (arnaudvergnet) [inline android/src/app/organicmaps/maplayer/MapButtonsController.java:377]: This function is not needed anymore, you can call `updateButtonsVisibility()` directly and it will handle the zoom buttons

File: android/src/app/organicmaps/maplayer/MapButtonsController.java [hunk 5]
```diff
@@ -272,6 +274,8 @@ public void showMapButtons(boolean show)
       UiUtils.show(mFrame);
     else
       UiUtils.hide(mFrame);
+
+    resumeZoomButtons();
```
Comment (arnaudvergnet) [inline android/src/app/organicmaps/maplayer/MapButtonsController.java:278]: As said in another comment, you can use `updateButtonsVisibility` directly. To improve performance it is also better to call this function only when `show` is true: we don't need to loop over all the buttons to detect if they should show or not if we want to hide them all

File: android/src/app/organicmaps/maplayer/MapButtonsController.java [hunk 6]
```diff
@@ -242,8 +242,10 @@ private void applyMove(View frame, float translationY)
 
   public void updateButtonsVisibility()
   {
-    updateButtonsVisibility(mInnerLeftButtonsFrame.getTranslationY(), mInnerLeftButtonsFrame);
-    updateButtonsVisibility(mInnerRightButtonsFrame.getTranslationY(), mInnerRightButtonsFrame);
+    if(mInnerRightButtonsFrame != null && mInnerLeftButtonsFrame != null) {
```
Comment (arnaudvergnet) [inline android/src/app/organicmaps/maplayer/MapButtonsController.java:245]: `mInnerRightButtonsFrame` could be null when `mInnerLeftButtonsFrame` is not (and also the opposite) depending on the current layout loaded.

So it would be better to check each UI frame before each associated function: only check `mInnerLeftButtonsFrame != null` before calling `updateButtonsVisibility(mInnerLeftButtonsFrame.getTranslationY...` and same for `mInnerRightButtonsFrame`.

**Review by arnaudvergnet — CHANGES_REQUESTED** [review]

**Comment by Aryan-Sagavekar** [comment]
I applied the changes you suggested, tested it too and it works fine.

**Comment by Jean-BaptisteC** [comment]
Can you please follow [Java style](https://github.com/organicmaps/organicmaps/blob/master/docs/JAVA_STYLE.md) and rebase your branch :)

**Comment by Aryan-Sagavekar** [comment]
Sure!

**Comment by arnaudvergnet** [comment]
We'll squash when merging so you don't need to bother. We keep the PR commit history only when it makes sense (for example one PR fixes 2 bugs, one fix in each commit).

When I work on PR I usually rebase on master regularly then force push the branch, instead of merging. This way we don't have merge commits in the PR and the history is easier to read. 

Regarding the code style, please try not to reformat code which is not related to your PR as this makes looking back at the history using `git blame` harder.

**Comment by Aryan-Sagavekar** [comment]
I'll reformat the code in my PR and rebase it again.

**Comment by Aryan-Sagavekar** [comment]
i guess i messed some things up while rebasing, i am so sorry for that. I'll just create a new PR with proper commits.

**Comment by arnaudvergnet** [comment]
please don't create a new pr, try to fix this one

**Comment by Aryan-Sagavekar** [comment]
I tried to fix it but it made it worse.

**Comment by arnaudvergnet** [comment]
Save the changes you want to include in the fix somewhere outside of the project so it is safe. Then recreate the `zoom_buttons` branch on your fork, and add a commit with your fix on it.

**Comment by arnaudvergnet** [comment]
If you are having trouble with git branches, I highly suggest going through https://learngitbranching.js.org

**Comment by Aryan-Sagavekar** [comment]
OK! Done i suppose. 
Thank you sir for guiding me through this one, I have learnt many things and I'm willing to continue to contribute to the project.

File: android/src/app/organicmaps/maplayer/MapButtonsController.java [hunk 7]
```diff
@@ -242,8 +241,14 @@ private void applyMove(View frame, float translationY)
...
-    updateButtonsVisibility(mInnerLeftButtonsFrame.getTranslationY(), mInnerLeftButtonsFrame);
-    updateButtonsVisibility(mInnerRightButtonsFrame.getTranslationY(), mInnerRightButtonsFrame);
+    if (mInnerLeftButtonsFrame != null)
+    {
+      updateButtonsVisibility(mInnerLeftButtonsFrame.getTranslationY(), mInnerLeftButtonsFrame);
+    }
+    if (mInnerRightButtonsFrame != null)
+    {
+      updateButtonsVisibility(mInnerRightButtonsFrame.getTranslationY(), mInnerRightButtonsFrame);
+    }
```
Comment (biodranik) [inline android/src/app/organicmaps/maplayer/MapButtonsController.java:251]:
```suggestion
      updateButtonsVisibility(mInnerLeftButtonsFrame.getTranslationY(), mInnerLeftButtonsFrame);
    if (mInnerRightButtonsFrame != null)
      updateButtonsVisibility(mInnerRightButtonsFrame.getTranslationY(), mInnerRightButtonsFrame);
```

**Comment by arnaudvergnet** [comment]
Nice, glad you managed to solve the problem! There's one change which got lost in the rebase: https://github.com/organicmaps/organicmaps/pull/4689/files#diff-cf6f2ac9b2276793755f87563f2ac026d4cd698f5b097d4a6fc4d5c6f1d3f403R262 the condition needs to be `>=`.

**Comment by Aryan-Sagavekar** [comment]
Oh right i'll fix it right away!

**Review by biodranik — APPROVED** [review]
Thank you, and thanks to @arnaudvergnet for the mentorship! Looks like we need to publish a manual for new contributors about our approach:
- Use git rebase and amend commits
- Do not close PRs
- Follow code guides
- Carefully study the code, corner cases, and test the fix
- Don't be afraid to go out of the box with simpler ideas and refactorings, and with a better UX
- ... something else?

**Comment by Aryan-Sagavekar** [comment]
Maybe adding references related to the code, so that new contributors can get a better understanding.

**Comment by arnaudvergnet** [comment]
@Aryan-Sagavekar we work on the project in our free time so PR can take some time before they get merged to let everyone have time to have a look.

**Comment by Jean-BaptisteC** [comment]
Problem is not full fixed:
- zoom buttons appears during routing if button zoom settings = false
- select POI during routing -> zoom buttons disappear and if leave routing zoom buttons is shown on main IHM
- select POI upate buttons visibility

**Comment by Aryan-Sagavekar** [comment]
adding the showmapbuttons() function to onResume() should solve this problem.
